EIPS/eip-custom_errors_for_erc_tokens.md

```diff
@@ -0,0 +1,393 @@
+
+Considering the outlined above, the errors are designed based on the standard actions that can be performed on each token, and the [subjects](#actions-and-subjects) involved.
+
+### Actions and subjects
+
+The main actions that can be performed within a token are:
+
+- **Transfer**: An operation in which a _sender_ moves any form of _balance_ to a _receiver_.
+- **Approval**: An operation in which an _approver_ grants any form of _approval_ to an _operator_.
+
+Whose subjects outlined are expected to be exhaustive to represent _what_ can go wrong in a transaction, which can be shown as an error by adding a [failure prefix](#error-prefixes).
+
+Note that `Token` or `TokenID` MUST NOT be a subject for the following reasons:
+
+1. The [Domain](#domain) is explicit enough to identify it as a token error.
+2. Although `TokenID` may identify an error, the token itself SHOULD NOT the root of the error, but can be additional information.
+
+Consider that, in practice, some of the subjects in a contract might be called different while they represent the same. These are listed below:
+
+- **Balance**: Represents the access to a certain amount of tokens for an address.
+  - EIP-721: Instead of _balance_, an address is _owner_ of a token.
+- **Approval**: Represents a permission granted to an operator.
+  - EIP-20: Instead of _approval_, an operator has _allowance_.
+- **Operator**: Represents an address that has received approval over third-party tokens.
+  - EIP-20: Instead of _operator_, an address with allowance is an _spender_.
+
+If a subject is called different on a particular token standard, the error MUST be consistent with the standard's naming convention.
+
+### Error prefixes
+
+An error prefix is what complements a subject, so it can be read as an error.
+Developers can think about an error prefix as the _why_ an error happened.
+
+A prefix can be `Invalid` for general incorrectness, or more specific like `Insufficient` for amounts (or lack of).
+
+### Domain
+
+Although these errors are considered to be expressive enough, each error's arguments may vary depending on the token domain, causing a `DeclarationError` if there are errors with the same name and different arguments.
+
+An example of this is:
+
+```solidity
+InsufficientApproval(address spender, uint256 allowance, uint256 needed);
+InsufficientApproval(address operator, uint256 tokenId);
+```
+
+For that reason, a domain prefix is proposed to avoid declaration clashing, which is the name of the ERC and its corresponding number appended at the beginning.
+
+Example:
+
+```solidity
+ERC20InsufficientApproval(address spender, uint256 allowance, uint256 needed);
+ERC721InsufficientApproval(address operator, uint256 tokenId);
+```
+
+A prefix is also RECOMMENDED for cases when there's
+
+### Arguments
+
+The selection of arguments depends on the subject involved, and it SHOULD follow the order presented below:
+
+1. _Who_ is involved with the error (an `address sender`)
```

Comment:
```suggestion
1. _Who_ is involved with the error (eg. `address sender`)
```

Reply:
I'd accept only if we agree an `address` is the only thing that can be there.

The rule is broken with `ERC1155InvalidArrayLength(uint256, uint256)`.

Reply:
What do you mean? "Eg." does not mean that what follows is the only option.

Re: array length, let's discuss that in the context of "who vs what". I think it's very clear in that example that the lengths are "what" and not "who"!